### PR TITLE
Allow Bulk Delete In Org Vault

### LINKED ISSUE #534: Missing multiselect in organization's vault
When one uses a filter in one's personal vault, there are select boxes to select multiple objects at once. There is even a "select all" feature.
When one switches to the organization's vault, those check boxes are not displayed and objects can only deleted one by one.

### PR DESCRIPTION
**Pull request by addisonbeck**
## Objective
Allow bulk selection & delete from organization vaults similar to what is already available in the personal vault

## Code Changes
- unhid the selection checkbox next to ciphers in the org vault
- added bulk selection functions to the org vault
- added bulk delete functions to the org vault
- extracted the bulk actions dropdown to a dedicated component
- refactored the personal vault component to use above dedicated component
- moved some shared code to jslib

## Related Issues
closes #534 
dependent on [jslib PR 130](https://github.com/bitwarden/jslib/pull/130)
dependent on [server PR 830](https://github.com/bitwarden/server/pull/830)

### CODE REVIEW

**Comment by kspearrin** [comment]
Actions that occur from the org admin area could be invoked by a regular user of the org or by an org admin/owner. If it is an org admin/owner we need different server APIs to validate their ability to make those actions. See the regular single cipher delete action as an example.

**Comment by addisonbeck** [comment]
added admin-specific calls

File: src/app/app.module.ts [hunk 1]
```diff
@@ -201,6 +201,7 @@ import localeSv from '@angular/common/locales/sv';
 import localeUk from '@angular/common/locales/uk';
 import localeZhCn from '@angular/common/locales/zh-Hans';
 import localeZhTw from '@angular/common/locales/zh-Hant';
+import { BulkActionsComponent } from './vault/bulk-actions.component';
```
Comment (kspearrin) [inline src/app/app.module.ts:204]: should be sorted with other component imports

File: src/app/organizations/vault/vault.component.ts [hunk 2]
```diff
@@ -272,15 +272,15 @@ export class VaultComponent implements OnInit, OnDestroy {
...
             await this.ciphersComponent.refresh();
         });
-        childComponent.onDeletedCipher.subscribe(async (c: CipherView) => {
+        childComponent.onDeletedCipher.subscribe(async () => {
             this.modal.close();
             await this.ciphersComponent.refresh();
         });
-        childComponent.onRestoredCipher.subscribe(async (c: CipherView) => {
+        childComponent.onRestoredCipher.subscribe(async () => {
             this.modal.close();
```
Comment (kspearrin) [inline src/app/organizations/vault/vault.component.ts:284]: Why was this changed? The signature for these outputs passes a CipherView:

```
    @Output() onSavedCipher = new EventEmitter<CipherView>();
    @Output() onDeletedCipher = new EventEmitter<CipherView>();
    @Output() onRestoredCipher = new EventEmitter<CipherView>();
    @Output() onCancelled = new EventEmitter<CipherView>();
    @Output() onEditAttachments = new EventEmitter<CipherView>();
    @Output() onShareCipher = new EventEmitter<CipherView>();
    @Output() onEditCollections = new EventEmitter<CipherView>();

```

File: src/app/vault/bulk-actions.component.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            {{'moveSelected' | i18n}}
+        </button>
+        <button class="dropdown-item" appStopClick (click)="bulkShare()" *ngIf="!deleted && !organization">
+            <i class="fa fa-fw fa-share-alt" aria-hidden="true"></i>
+            {{'shareSelected' | i18n}}
+        </button>
+        <button class="dropdown-item" (click)="bulkRestore()" *ngIf="deleted && !organization">
+            <i class="fa fa-fw fa-undo" aria-hidden="true"></i>
+            {{'restoreSelected' | i18n}}
+        </button>
```
Comment (kspearrin) [inline src/app/vault/bulk-actions.component.html:18]: I am guessing we would probably want restore bulk action too, right @cscharf ? Maybe a fast follow?

File: src/app/vault/bulk-actions.component.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
+import {
+    Component, ComponentFactoryResolver, Input, ViewChild, ViewContainerRef,
```
Comment (kspearrin) [inline src/app/vault/bulk-actions.component.ts:2]: Existing formatting always new lines these

File: src/app/vault/bulk-delete.component.ts [hunk 5]
```diff
@@ -4,12 +4,13 @@ import {
     Input,
     Output,
 } from '@angular/core';
-
 import { ToasterService } from 'angular2-toaster';
 import { Angulartics2 } from 'angulartics2';
-
+import { ApiService } from 'jslib/abstractions';
```
Comment (kspearrin) [inline src/app/vault/bulk-delete.component.ts:9]: import from `jslib/abstractions/api.service` like others

File: src/app/vault/bulk-delete.component.ts [hunk 6]
```diff
@@ -18,17 +19,25 @@ import { I18nService } from 'jslib/abstractions/i18n.service';
...
     @Input() cipherIds: string[] = [];
     @Input() permanent: boolean = false;
+    @Input() organization: Organization;
     @Output() onDeleted = new EventEmitter();
 
-    formPromise: Promise<any>;
-
     constructor(private analytics: Angulartics2, private cipherService: CipherService,
-        private toasterService: ToasterService, private i18nService: I18nService) { }
+        private toasterService: ToasterService, private i18nService: I18nService, private apiService: ApiService) { }
```
Comment (kspearrin) [inline src/app/vault/bulk-delete.component.ts:26]: two per line

File: src/app/vault/bulk-delete.component.ts [hunk 7]
```diff
@@ -18,17 +19,25 @@ import { I18nService } from 'jslib/abstractions/i18n.service';
...
+        if (!this.organization || !this.organization.isAdmin) {
+            this.permanent
+                ? await this.cipherService.deleteManyWithServer(this.cipherIds)
+                : await this.cipherService.softDeleteManyWithServer(this.cipherIds);
+        } else {
+            this.permanent
+                ? await this.apiService
+                    .deleteManyCiphersAdmin(new CipherBulkDeleteRequest(this.cipherIds, this.organization.id))
+                : await this.apiService
+                    .putDeleteManyCiphersAdmin(new CipherBulkDeleteRequest(this.cipherIds, this.organization.id));
```
Comment (kspearrin) [inline src/app/vault/bulk-delete.component.ts:38]: I think long-hand if/else would probably read better here

File: src/app/vault/vault.component.ts [hunk 8]
```diff
@@ -83,7 +77,7 @@ export class VaultComponent implements OnInit, OnDestroy {
         private i18nService: I18nService, private componentFactoryResolver: ComponentFactoryResolver,
         private tokenService: TokenService, private cryptoService: CryptoService,
         private messagingService: MessagingService, private userService: UserService,
-        private platformUtilsService: PlatformUtilsService, private toasterService: ToasterService,
+        private platformUtilsService: PlatformUtilsService,
         private broadcasterService: BroadcasterService, private ngZone: NgZone) { }
```
Comment (kspearrin) [inline src/app/vault/vault.component.ts:81]: keep two per line until the end line

File: src/app/vault/vault.component.ts [hunk 9]
```diff
@@ -366,15 +360,15 @@ export class VaultComponent implements OnInit, OnDestroy {
...
             await this.ciphersComponent.refresh();
         });
-        childComponent.onDeletedCipher.subscribe(async (c: CipherView) => {
+        childComponent.onDeletedCipher.subscribe(async () => {
             this.modal.close();
             await this.ciphersComponent.refresh();
         });
-        childComponent.onRestoredCipher.subscribe(async (c: CipherView) => {
+        childComponent.onRestoredCipher.subscribe(async () => {
             this.modal.close();
```
Comment (kspearrin) [inline src/app/vault/vault.component.ts:372]: Same as above. Why did this change?

File: src/app/organizations/vault/vault.component.ts (hunk 2 above)
Comment (addisonbeck) [reply src/app/organizations/vault/vault.component.ts:284]: This was just a linter recommendation, since the value is never used, but I could see a reason for keeping it around for context.

File: src/app/vault/bulk-actions.component.html (hunk 3 above)
Comment (addisonbeck) [reply src/app/vault/bulk-actions.component.html:18]: There is a backlog item for other bulk items, including restore. I imagine I'll grab that as soon as it's planned and our August stuff is all wrapped up.

File: src/app/app.module.ts (hunk 1 above)
Comment (addisonbeck) [reply src/app/app.module.ts]: Done

File: src/app/vault/bulk-actions.component.ts (hunk 4 above)
Comment (addisonbeck) [reply src/app/vault/bulk-actions.component.ts]: Fixed

File: src/app/vault/bulk-delete.component.ts (hunk 6 above)
Comment (addisonbeck) [reply src/app/vault/bulk-delete.component.ts]: Done

File: src/app/vault/bulk-delete.component.ts (hunk 7 above)
Comment (addisonbeck) [reply src/app/vault/bulk-delete.component.ts]: Kinda done, I created some private methods that use long hand if/else and called those.

File: src/app/vault/vault.component.ts (hunk 8 above)
Comment (addisonbeck) [reply src/app/vault/vault.component.ts]: Done

File: src/app/organizations/vault/vault.component.ts (hunk 2 above)
Comment (addisonbeck) [reply src/app/organizations/vault/vault.component.ts:284]: Let me know if you still want this put back

File: src/app/vault/vault.component.ts (hunk 9 above)
Comment (addisonbeck) [reply src/app/vault/vault.component.ts:372]: Responded above

File: src/app/vault/bulk-delete.component.ts (hunk 5 above)
Comment (addisonbeck) [reply src/app/vault/bulk-delete.component.ts]: Done

File: src/app/organizations/vault/vault.component.ts (hunk 2 above)
Comment (cscharf) [reply src/app/organizations/vault/vault.component.ts:284]: Please do, thanks @addisonbeck

Comment (addisonbeck) [reply src/app/organizations/vault/vault.component.ts]: It is done 😁

File: src/app/vault/bulk-delete.component.ts [hunk 10]
```diff
@@ -4,12 +4,13 @@ import {
...
     Output,
 } from '@angular/core';
-
 import { ToasterService } from 'angular2-toaster';
 import { Angulartics2 } from 'angulartics2';
-
+import { ApiService } from 'jslib/abstractions/api.service';
 import { CipherService } from 'jslib/abstractions/cipher.service';
 import { I18nService } from 'jslib/abstractions/i18n.service';
+import { Organization } from 'jslib/models/domain/organization';
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/app/vault/bulk-delete.component.ts:14]: Done! Separated abstractions too, since that seems to be the standard.

File: src/app/vault/bulk-actions.component.html [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            <i class="fa fa-fw fa-minus-square-o" aria-hidden="true"></i>
+            {{'unselectAll' | i18n}}
+        </button>
+    </div>
+</div>
+
+<ng-template #bulkDeleteTemplate></ng-template>
+<ng-template #bulkRestoreTemplate></ng-template>
+<ng-template #bulkMoveTemplate></ng-template>
+<ng-template #bulkShareTemplate></ng-template>
```
Comment (kspearrin) [inline src/app/vault/bulk-actions.component.html:38]: Don't you need to remove these from other areas now? See https://github.com/bitwarden/web/pull/577/files#diff-867123f4cf0e9410557f5fd01cd28952R93-R96

Comment (addisonbeck) [reply src/app/vault/bulk-actions.component.html:38]: Done

File: src/app/vault/ciphers.component.ts [hunk 12]
```diff
@@ -38,6 +36,8 @@ export class CiphersComponent extends BaseCiphersComponent implements OnDestroy
     cipherType = CipherType;
     actionPromise: Promise<any>;
 
+    private maxCheckedCount = 500;
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/app/vault/ciphers.component.ts]: From when it briefly was moved to jslib. Fixed.

File: src/app/vault/bulk-actions.component.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+import { ToasterService } from 'angular2-toaster';
+import { I18nService } from 'jslib/abstractions/i18n.service';
+import { Organization } from 'jslib/models/domain/organization';
+import { ModalComponent } from '../modal.component';
+import { BulkDeleteComponent } from './bulk-delete.component';
+import { BulkMoveComponent } from './bulk-move.component';
+import { BulkRestoreComponent } from './bulk-restore.component';
+import { BulkShareComponent } from './bulk-share.component';
+import { CiphersComponent } from './ciphers.component';
+
```
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/app/vault/bulk-actions.component.ts]: Done

**Comment by kspearrin** [comment]
Looks like you need to resolve some jslib conflicts.

**Comment by addisonbeck** [comment]
@kspearrin @cscharf can one of you check this out one last time? Doing a pull/rebase to resolve jslib conflicts dismissed the approval.

**Comment by addisonbeck** [comment]
@cscharf one more time please, those breaking changes from jslib caught me off guard 😰
